assembly/src/assembler/mod.rs

```diff
@@ -291,6 +293,25 @@ impl Assembler {
 
     // PROCEDURE GETTER
     // --------------------------------------------------------------------------------------------
+
+    /// Return a procedure if it exists in the compiled procedures cache.
+    ///
+    /// This function will create a borrow from the refcell of the procedure cache. Any mutable
+    /// access to the inner proc cache must be checked; otherwise, this function might panic
+    unsafe fn get_imported_proc_without_compiling(
+        &self,
+        proc_id: &ProcedureId,
+    ) -> Option<Ref<'_, Procedure>> {
+        let proc_cache = self.proc_cache.borrow();
+
+        // if the procedure is already in the procedure cache, return it
+        if proc_cache.contains_key(proc_id) {
+            Some(Ref::map(proc_cache, |cache| cache.get(proc_id).unwrap()))
+        } else {
+            None
+        }
+    }
```

Comment:
I wonder if we could avoid `unsafe` code completely. For example, what if instead of this function we have something like this:
```Rust
fn ensure_procedure_is_in_cache(
    &self,
    proc_id: &ProcedureId,
    context: &mut AssemblyContext,
) -> Result<(), AssemblyError> {
    if !self.proc_cache.borrow().contains_key(proc_id) {
        // if procedure is not in cache, try to get its module and compile it
        let module = self
            .module_provider
            .get_module(proc_id)
            .ok_or_else(|| AssemblyError::imported_proc_module_not_found(proc_id))?;
        self.compile_module(&module, module.path(), context)?;

        // if the procedure is still not in cache, then there was some error
        if !self.proc_cache.borrow().contains_key(proc_id) {
            return Err(AssemblyError::imported_proc_not_found_in_module(
                proc_id,
                module.path(),
            ));
        }
    }

    Ok(())
}
```
Then, we could change code in `exec_imported()` (and `call_imported()`) to something like this:
```Rust
pub(super) fn exec_imported(
    &self,
    proc_id: &ProcedureId,
    context: &mut AssemblyContext,
) -> Result<Option<CodeBlock>, AssemblyError> {
    // make sure the procedure is in procedure cache
    self.ensure_procedure_is_in_cache(proc_id, context)?;

    // get the procedure from the assembler
    let proc_cache = self.proc_cache.borrow();
    let proc = proc_cache.get(proc_id).expect("procedure not in cache");
    debug_assert!(proc.is_export(), "not imported procedure");

    // register and "inlined" call to the procedure; this updates the callset of the
    // procedure currently being compiled
    context.register_external_call(proc, true)?;

    // TODO: if the procedure consists of a single SPAN block, we could just append all
    // operations from that SPAN block to the span builder instead of returning a code block

    // return the code block of the procedure
    Ok(Some(proc.code_root().clone()))
}
```
If we do this, we should also be able to get rid of `get_imported_proc()` method.

Reply:
I'm not sure if that will work because we `self.compile_module(&module, module.path(), context)?;` inside of the borrow. Theoretically, we can't attempt to mut access the refcell.

That unsafe directive is just to warn the developer (not user, because it is not exposed by the API) to not attempt to borrow mut while using the function. However, unsafe is reserved only for undefined memory behavior, so we can just completely remove that. I think we will still have to keep the separation between borrow/mut

Reply:
I think it should work as borrow in `if !self.proc_cache.borrow().contains_key(proc_id) {` should be released before we go into `compile_module()`. Could you check this?

Reply:
That's true, we return an owned `bool` so it should release it. Will check now!

Reply:
It works fine! What if instead of repeating

```
// get the procedure from the assembler
    let proc_cache = self.proc_cache.borrow();
    let proc = proc_cache.get(proc_id).expect("procedure not in cache");
    debug_assert!(proc.is_export(), "not imported procedure");
```

We just keep the 

```
    fn get_imported_proc(
        &self,
        proc_id: &ProcedureId,
        context: &mut AssemblyContext,
    ) -> Result<Ref<'_, Procedure>, AssemblyError> {
```

So we centralize in a single place the logic to extract the procedure from the cache?

Reply:
Another change I think its better is if we request the mutable borrow of the proc cache in the arguments of `fn compile_module`. That way, we guarantee the caller took care to safely borrow this mut.

We currently do that only in two places: when we are loading the kernel, and when we get an imported proc

Reply:
I made a slightly different version from your suggestion so we don't expose the need to borrow/unwrap from the proc cache. Let me know wdyt

Reply:
Thank you! I mention this in the comment below, but, at least for me, the approach with `ensure_procedure_is_in_cache()` seems simpler to follow. So, I'd prefer to go with that approach.